### PR TITLE
drivers:can: Fix R-Car driver can_stop behaviour

### LINKED ISSUE #50546: drivers: can: rcar: inconsistent behavior when calling can_stop() with pending transmissions
**Describe the bug**
The Renesas R-Car CAN controller drivers likely exhibits inconsistent behavior when calling `can_stop()` with pending CAN frame transmissions similar as to what was reported in #50545. I have not been able to find the behavior of the Renesas R-Car CAN driver as I do not have access to the hardware.

**To check**
1. Disconnect the CAN controller from the CAN bus.
2. Using the CAN shell, call `can start <device>`.
3. Using the CAN shell, call `can send <device> 010`.
4. Using the CAN shell, call `can stop <device>`.
5. Connect the CAN controller to a CAN bus with at least one other node for ACK'ing frames and monitor any received frames on that node (e.g. use `candump` on GNU/Linux).
6. Using the CAN shell, call `can start <device>`.
7. Observe the different behavior across the Zephyr CAN drivers as described above.

**Expected behavior**
The behavior should be the same across the Zephyr CAN drivers. Since some CAN controllers automatically abort any pending transmissions, the common behavior should be to abort any pending transmissions and notify the senders of this.

**Impact**
Unknown.

**Environment (please complete the following information):**
 - Commit SHA: 7e6e4579c29f269626d52bb70d5eed30113d90ee

**Additional context**
The `can_stop()` API was introduced in #50102.

### PR DESCRIPTION
**Pull request by aaillet**
Calling for can_stop() method on R-Car boards was not emptying FIFOs and tx_msgq.

We use controller reset mode to empty controller FIFOs.
We are emptying tx_msgq with "-ENETDOWN" return code.

Tested using can shell module ans obtained the following behaviour:
```
#Board disconnected from CAN bus
uart:~$ can start can@e6c30000
starting can@e6c30000
uart:~$ can send can@e6c30000 001 DE AD BE EF
enqueuing CAN frame #1 with standard (11-bit) CAN ID 0x001, RTR 0, CAN-FD 0, BRS 0, DLC 4
uart:~$ can send can@e6c30000 001 DE AD BE
enqueuing CAN frame #2 with standard (11-bit) CAN ID 0x001, RTR 0, CAN-FD 0, BRS 0, DLC 3
uart:~$ can send can@e6c30000 001 DE AD
enqueuing CAN frame #3 with standard (11-bit) CAN ID 0x001, RTR 0, CAN-FD 0, BRS 0, DLC 2
uart:~$ can stop can@e6c30000
stopping can@e6c30000
failed to send CAN frame #1 (err 141)
failed to send CAN frame #2 (err 141)
failed to send CAN frame #3 (err 141)
#Connect board to CAN bus
uart:~$ can start can@e6c30000
starting can@e6c30000
uart:~$ can send can@e6c30000 001 DE AD BE EF
enqueuing CAN frame #4 with standard (11-bit) CAN ID 0x001, RTR 0, CAN-FD 0, BRS 0, DLC 4
CAN frame #4 successfully sent
```
First three frames has been properly dropped and the last one has been properly sent.

Fix #50546

### CODE REVIEW

File: drivers/can/can_rcar.c [hunk 1]
```diff
@@ -623,9 +623,9 @@ static int can_rcar_stop(const struct device *dev)
 
 	k_mutex_lock(&data->inst_mutex, K_FOREVER);
 
-	ret = can_rcar_enter_halt_mode(config);
+	ret = can_rcar_enter_reset_mode(config, true);
```
Comment (henrikbrixandersen) [inline drivers/can/can_rcar.c:626]: Are the installed RX filters preserved when entering reset mode (does the `tests/drivers/can/api` test suite pass with this change)?

**Comment by henrikbrixandersen** [comment]
@aaillet ping

**Comment by aaillet** [comment]
> @aaillet ping

I haven't forgotten, I'm on vacation, I'll take care of it tomorrow !

**Comment by aaillet** [comment]
So you were right, RX filters weren't kept with the changes I proposed (dropped by resetting controller). 
Instead of switching to reset mode I'm now switching to halt mode and manually resetting RX FIFO.

My changes weren't passing can api test suite, new version is passing the test suite as before.
By testing the driver manually using can shell, I can confirm we are:
- Dropping pending transmission, emptying RX FIFO and returning an error for each message in RX msg (ENETDOWN @henrikbrixandersen can you confirm it's the right error code to return ?) when stopping controller
- Keeping RX filters when restarting controller (we are also receiving frames received when controller was stopped)
- Not altering controller config as we are still able to send message after a controller restart (stop+start)

```
*** Booting Zephyr OS build zephyr-v3.3.0-2972-g3a08d9620ae5 ***
Starting Zephyr shell. (rcar_h3ulcb_cr7)
# Controller unlinked from can bus
uart:~$ can start can@e6c30000 
starting can@e6c30000
uart:~$ can send can@e6c30000 001 DE
enqueuing CAN frame #0 with standard (11-bit) CAN ID 0x001, RTR 0, CAN-FD 0, BRS 0, DLC 1
uart:~$ can filter add can@e6c30000 001
adding filter with standard (11-bit) CAN ID 0x001, CAN ID mask 0x7ff, data frames 1, RTR frames 0, CAN-FD frames 0
filter ID: 0
uart:~$ can stop can@e6c30000 
stopping can@e6c30000
failed to send CAN frame #0 (err 141)
# Link controller to can bus
uart:~$ can start can@e6c30000
starting can@e6c30000
uart:~$
     001   [1]  de 
uart:~$ can send can@e6c30000 001 DE
enqueuing CAN frame #1 with standard (11-bit) CAN ID 0x001, RTR 0, CAN-FD 0, BRS 0, DLC 1
CAN frame #1 successfully sent
uart:~$ can stop can@e6c30000 
stopping can@e6c30000
# Sending frames from Linux host while controller stopped
uart:~$ can start can@e6c30000
starting can@e6c30000
     001   [1]  de 
     001   [2]  de ad 
     001   [3]  de ad be 
     001   [4]  de ad be ef
uart:~$
# Sending frames from Linux host while controller started
     001   [5]  de ad be ef de 
     001   [6]  de ad be ef de ad 
```

Only thing I don't understand is why frames are dropped with error code 141 while `-ENETDOWN` is error 100, as can shell is directly reporting error from can api `can_send`, I don't see where the translation is done.

**Comment by henrikbrixandersen** [comment]
> * Keeping RX filters when restarting controller (we are also receiving frames received when controller was stopped)

This doesn't sound right. That likely means the controller also sent ACKs for those frames while stopped? After calling `can_stop()`, the CAN controller must not interact with the network.

**Comment by henrikbrixandersen** [comment]
> * Dropping pending transmission, emptying RX FIFO and returning an error for each message in RX msg (ENETDOWN @henrikbrixandersen can you confirm it's the right error code to return ?) when stopping controller

Yes, `-ENETDOWN` is correct in this case.

**Comment by henrikbrixandersen** [comment]
> > * Keeping RX filters when restarting controller (we are also receiving frames received when controller was stopped)
> 
> This doesn't sound right. That likely means the controller also sent ACKs for those frames while stopped? After calling `can_stop()`, the CAN controller must not interact with the network.

@aaillet Perhaps re-use what is done in `can_rcar_set_timing()` where registers are manually restored after entering/existing reset mode?

**Comment by aaillet** [comment]
> > * Keeping RX filters when restarting controller (we are also receiving frames received when controller was stopped)
> 
> This doesn't sound right. That likely means the controller also sent ACKs for those frames while stopped? After calling `can_stop()`, the CAN controller must not interact with the network.

I'll investigate on what is going on, I'm surprised that even in halt mode, the controller is ACKing frames.

**Comment by henrikbrixandersen** [comment]
> I'll investigate on what is going on, I'm surprised that even in halt mode, the controller is ACKing frames.

@aaillet Ah, I think I know what is going on. When in "halt" after having called can_stop(), the remote node attempts to transmit a frame but receives no ACK. The remote node continuously attempts to retransmit the frame and once the local node exits "halt" after calling can_start(), it receives and ACKs the retransmission.

If you can confirm this behaviour (that the remote node does not receive any ACKs when calling can_stop() on the local node, using "halt" mode should be fine.

**Comment by henrikbrixandersen** [comment]
> > I'll investigate on what is going on, I'm surprised that even in halt mode, the controller is ACKing frames.
> 
> @aaillet Ah, I think I know what is going on. When in "halt" after having called can_stop(), the remote node attempts to transmit a frame but receives no ACK. The remote node continuously attempts to retransmit the frame and once the local node exits "halt" after calling can_start(), it receives and ACKs the retransmission.
> 
> If you can confirm this behaviour (that the remote node does not receive any ACKs when calling can_stop() on the local node, using "halt" mode should be fine.

@aaillet Can you confirm the above behaviour? It would be nice to have this fix in for v.3.4.0.

**Comment by aaillet** [comment]
> > > I'll investigate on what is going on, I'm surprised that even in halt mode, the controller is ACKing frames.
> > 
> > 
> > @aaillet Ah, I think I know what is going on. When in "halt" after having called can_stop(), the remote node attempts to transmit a frame but receives no ACK. The remote node continuously attempts to retransmit the frame and once the local node exits "halt" after calling can_start(), it receives and ACKs the retransmission.
> > If you can confirm this behaviour (that the remote node does not receive any ACKs when calling can_stop() on the local node, using "halt" mode should be fine.
> 
> @aaillet Can you confirm the above behaviour? It would be nice to have this fix in for v.3.4.0.

Yes, I need to figure out how to know if my remote node (linux using cansend with a canusb module) received the ack.
I'll get back to you ASAP !

**Comment by aaillet** [comment]
Hey @henrikbrixandersen,
I just validated that when controller is in halt mode (after calling for `can_stop`) it doesn't send any acknowledgment to my remote node.
I also validated that my remote node is "repeatedly sending the frame on the CAN bus until there is an acknowledgement from at least one other node."
I can then confirm the behavior you were thinking about !
I do think now that my patch is correctly fixing `can_stop` behavior. (still passing can api test suite and working well when tested using can shell).

**Review by henrikbrixandersen — APPROVED** [review]
Thanks!

**Review by pmarzin — APPROVED** [review]
LGTM :) Thanks @henrikbrixandersen and @aaillet !

**Pull request merged**